[Project1/김유경] Notion 클로닝 프로젝트

kimyou1102 opened the pull request:
## 📌 과제 설명
바닐라JS를 이용해 노션 클로닝 프로젝트를 수행하였습니다.

## 👩‍💻 요구 사항과 구현 내용
![image](https://github.com/prgrms-fe-devcourse/FEDC5-5_Project_Notion_VanillaJS/assets/66080362/56d1bfb6-7522-47d5-bc10-8a2b2901593c)

### 기본 요구 사항
- [x] 바닐라 JS만을 이용해 노션을 클로닝합니다.
- [x] 기본적인 레이아웃은 노션과 같으며, 스타일링, 컬러값 등은 원하는대로 커스텀합니다.
- [x] 글 단위를 Document라고 합니다. Document는 Document 여러개를 포함할 수 있습니다.
- [x] 화면 좌측에 Root Documents를 불러오는 API를 통해 루트 Documents를 렌더링합니다.
- [x] Root Document를 클릭하면 오른쪽 편집기 영역에 해당 Document의 Content를 렌더링합니다.
- [x] 해당 Root Document에 하위 Document가 있는 경우, 해당 Document 아래에 트리 형태로 렌더링 합니다.
- [x] Document Tree에서 각 Document 우측에는 + 버튼이 있습니다. 해당 버튼을 클릭하면, 클릭한 Document의 하위 Document로 새 Document를 생성하고 편집화면으로 넘깁니다.
- [x] 편집기에는 기본적으로 저장 버튼이 없습니다. Document Save API를 이용해 지속적으로 서버에 저장되도록 합니다.
- [x] History API를 이용해 SPA 형태로 만듭니다.
- [x] 루트 URL 접속 시엔 별다른 편집기 선택이 안 된 상태입니다.
- [x] /documents/{documentId} 로 접속시, 해당 Document 의 content를 불러와 편집기에 로딩합니다.

### 보너스 요구사항
- [x] 기본적으로 편집기는 textarea 기반으로 단순한 텍스트 편집기로 시작하되, 여력이 되면 div와 contentEditable을 조합해서 좀 더 Rich한 에디터를 만들어봅니다.
- [ ] 편집기 최하단에는 현재 편집 중인 Document의 하위 Document 링크를 렌더링하도록 추가합니다.
- [ ] 편집기 내에서 다른 Document name을 적은 경우, 자동으로 해당 Document의 편집 페이지로 이동하는 링크를 거는 기능을 추가합니다.

### 구현 중 문제점 
- [x] 커서 위치가 이상한 곳으로 가는 문제
- [ ] 삭선, 볼드, 이탤릭을 여러개 적용시키지 못하는 문제
- [ ] 입력중에 api요청이 되면 마지막 글자가 반영 안되는 문제

## ✅ 피드백 반영사항
- [x] 3개이상의 if-else문 switch문으로 변경
- [x] 선언부와 가깝게 변수 위치 변경
- [x] 불필요한 코드 제거 및 오타 수정
- [x] history.replaceState기능의 함수화
- [ ] DocumentListItem 컴포넌트화

## ✅ PR 포인트 & 궁금한 점
- 기능이 많아지면서 커밋메시지가 제대로 쪼개지지 못한 거 같은데 커밋 메시지를 더 잘게 쪼개야할지 아니면 묶어서 해도되는지 궁금합니다.
- 컴포넌트가 서로 의존성이 낮도록 잘 분리되었는지 궁금합니다.
- contentEditable활용하면서 커서 문제를 아직 해결하지 못했는데 setState를 없애서 하는 것이 나을까요, 아니면 setState로 변경될때마다 커서를 찾아주는 것이 맞을까요?
- contentEditable에서 innerHTML이 변경될 때마다 정확한 위치에 커서를 둘 수 있는 방법이 궁금합니다.

kimyou1102 committed:
feat: 더미데이터로 문서 트리 ui 구현

kimyou1102 committed:
feat: 문서 트리 렌더 api연동

kimyou1102 committed:
feat: 편집기 생성

kimyou1102 committed:
feat: routing 처리 구현

kimyou1102 committed:
feat: route 커스텀 이벤트 생성

kimyou1102 committed:
feat: 문서이름 클릭 시 해당 문서의 편집기로 이동 구현

kimyou1102 committed:
feat: 하위 document추가 기능 구현

kimyou1102 committed:
feat: 문서 삭제 기능 구현

kimyou1102 committed:
fix: 현재 문서로 또 이동 시 편집기가 비게되는 문제 해결

kimyou1102 committed:
fix: 새 문서 생성 시 제목만 저장되는 오류 해결

kimyou1102 committed:
refactor: state를 App컴포넌트에서 하위에 내려주는 방식으로 변경

kimyou1102 committed:
feat: 문서 생성과 동시에 문서 리스트 업데이트

kimyou1102 committed:
refactor:  div와 contentEditable을 조합한 리치에디터로 변경

kimyou1102 committed:
style: css스타일 작업 및 아이콘 삽입

kimyou1102 committed:
feat: 글자 드래그로 선택 시 편집버튼 보여주는 기능 구현

kimyou1102 committed:
feat: 뒤로가기 앞으로가기 기능 추가

kimyou1102 committed:
refactor: 문자열 변환 정규식으로 변경

mojoblue7 commented on src/components/DocumentList.js:
```diff
@@ -0,0 +1,111 @@
+    documents.forEach((document) => {
+
+      htmlString += `
+        <div class="documents-item">
+            <div class="document-item-title-wrap ${document.id === parseInt(this.state.documentId) ? 'selected-document' : ''}" data-id="${document.id}">
+                <button class="toggle">
+                  <i class="fas fa-chevron-right" style="color: #797979;"></i>
+                </button>
+                <span class="document-item-title">${document.title}</span>
+                <div class="action-buttons">
+                  <button class="remove-document">
+                    <i class="fas fa-trash-alt" style="color: #797979;"></i>
+                  </button>
+                  <button class="add-document">
+                    <i class="fas fa-plus" style="color: #797979;"></i>
+                  </button>
+                </div>
+                
+            </div>
+            <div class="documents-item fold">
+                ${
+                  document.documents && document.documents.length !== 0
+                    ? documentPrint(document.documents)
+                    : '<span class="end-document-item">하위 페이지 없음</span>'
+                }
+            </div>
+        </div>
+    `;
```
해당 코드를 DocumentListItem.js 와 같이 컴포넌트화 시켜서 단순화 시키는 것이 좋아보입니다 😃

kimyou1102 replied:
넵 수정사항들 반영해서 pr하겠습니다!!😀

mojoblue7 reviewed (commented):
Q1 ) 기능이 많아지면서 커밋메시지가 제대로 쪼개지지 못한 거 같은데 커밋 메시지를 더 잘게 쪼개야할지 아니면 묶어서 해도되는지 궁금합니다.
- 이번 프로젝트 커밋을 보면 적당하게 나누셨습니다. 커밋의 단위는 기능 단위나, 컴포넌트 단위로 하는 경우가 많습니다. 


Q2 ) 컴포넌트가 서로 의존성이 낮도록 잘 분리되었는지 궁금합니다.
- 코드 리뷰에 반영해 놓았습니다 😃 


Q3) contentEditable활용하면서 커서 문제를 아직 해결하지 못했는데 setState를 없애서 하는 것이 나을까요, 아니면 setState로 변경될때마다 커서를 찾아주는 것이 맞을까요?
- 글로만 봐서 어떤 문제인지 파악을 하지 못했습니다. 커피챗때, 코드를 보면서 한번 보면 좋을 것 같습니다 😃 

Q4) contentEditable에서 innerHTML이 변경될 때마다 정확한 위치에 커서를 둘 수 있는 방법이 궁금합니다.
- innerHTML을 변경하는 방식 말고 classList에 class를 추가 제거하는 작업으로 처리하면 커서 문제를 해결하면서 조금 더 쉽게 구현할 수 있을 것이라 추측됩니다. 😃

bel1c10ud commented on src/components/App.js:
```diff
@@ -0,0 +1,105 @@
+export default function App({ $target, initialState }) {
+  this.state = initialState;
+
+  this.setState = async (nextState) => {
+    if (nextState.documentId !== this.state.documentId) {
+      this.state = nextState;
+      const { documentId, selectedDocument } = this.state;
+
+      documentEditSection.setState({
+        documentId,
+        document: selectedDocument,
+      });
+    } else {
+      this.state = nextState;
+      const { documentId, documents } = this.state;
+
+      documentList.setState({
+        documentId,
+        documents,
+      });
+    }
+
+    this.render();
```
이런 식으로 중복 코드를 줄여볼 수 있을 것 같아요. 🙂
```js
this.setState = (nextState) => {
    const { documentId, selectedDocument, documents } = this.state;

    this.state = nextState;

    if (nextState.documentId !== this.state.documentId) {
      documentEditSection.setState({
        documentId,
        document: selectedDocument,
      });
    } else {
      documentList.setState({
        documentId,
        documents,
      });
    }

    this.render();
  };
```

bel1c10ud commented on src/components/DocumentEditSection.js:
```diff
@@ -0,0 +1,156 @@
+            }),
+          });
+
+          history.replaceState(null, null, `/document/${createDocument.id}`);
```
`history.replaceState`도 `push`처럼 `router.js`에 선언하고 사용하면 좋을 것 같아요

kimyou1102 replied:
저도 종운님 코드를 보면서 replace에 관한 함수도 따로 만들어서 관리하는 것이 좋을 거 같다고 생각했습니다! 반영해서 pr하겠습니다🙂

bel1c10ud commented on src/components/DocumentEditSection.js:
```diff
@@ -0,0 +1,156 @@
+          });
+
+          history.replaceState(null, null, `/document/${createDocument.id}`);
+          createDocument.content = getItem(documentLocalSaveKey, "").content;
```
혹시나 로컬 스토리지에서 가져오기를 실패하면 `defaultValue`가 문자열이라 오류가 발생할 것 같아요 🤔

kimyou1102 replied:
''가 아닌 state에 맞는 기본값으로 defaultValue를 해줘야겠네요.😅 놓쳤던 부분 알게 해주셔서 감사합니다:)

bel1c10ud commented on src/components/EditButtons.js:
```diff
@@ -0,0 +1,69 @@
+  $div.setAttribute("class", "hide edit-button-wrap");
+  $target.appendChild($div);
+
+  $div.innerHTML = `
+    <button class="bold-button">
+        <i class="fas fa-bold"></i>
+    </bold>
+    <button class="italic-button">
+        <i class="fas fa-italic"></i>
+    </bold>
+    <button class="strike-button">
+        <i class="fas fa-strikethrough"></i>
+    </bold>
+  `;
```
마크업에 오타가 있는 것 같아요. 🙂

bel1c10ud reviewed (commented):
전반적으로 컴포넌트나 함수, 변수 명칭이 의미를 이해하기 좋게 잘 작성된 것 같아요. 👍
고생 많으셨습니다.

kimyou1102 committed:
refactor: 에디터 커서 버그 수정

ienrum commented on src/components/App.js:
```diff
@@ -0,0 +1,105 @@
+export default function App({ $target, initialState }) {
+  this.state = initialState;
+
+  this.setState = async (nextState) => {
```
async 키워드를 뺄수도 있을것 같아요😳

ienrum commented on src/components/App.js:
```diff
@@ -0,0 +1,105 @@
+    },
+  });
+
+  const $editorSection = document.createElement("div");
+
+  this.render = () => {
+    $target.appendChild($editorSection);
+    $editorSection.setAttribute("class", "editor-page");
+  };
+
+  const documentEditSection = new DocumentEditSection({
+    $target: $editorSection,
```
일관 되게 태그 생성로직을 컴포넌트 안에서 해도 좋을것 같아요!😌

ienrum commented on src/components/DocumentEditSection.js:
```diff
@@ -0,0 +1,156 @@
+
+  this.state = initialState;
+
+  let documentLocalSaveKey = `temp-post-${this.state.documentId}`;
```
documentLocalSaveKey 가 이 컴포넌트에서 변경되면서 많이쓰여서 알아보기가 힘든것 같아요! 아니면 각 로직에서 필요하면 각각 코드 블록에서 알맞는 변수명을 생성해도 좋을것같아요..😀

kimyou1102 replied:
저도 짜면서 documentLocalSaveKey 많이 바뀐 거 같다고 인지했는데 각각의 코드 블록에서 생성하는 것이 훨씬 가독성이 좋을 거 같네요! pr에 반영하겠습니다.😀

ienrum commented on src/components/DocumentEditSection.js:
```diff
@@ -0,0 +1,156 @@
+      documentId: this.state.documentId,
+      document: this.state.document,
+    },
+    onEditing: (document) => {
```
변수명이 document 라서 전역 document 변수와 혼동할수도있을것 같아요..? 실수로 오염될 가능성도 있을것 같아요..🥲

ienrum commented on src/components/DocumentEditSection.js:
```diff
@@ -0,0 +1,156 @@
+          history.replaceState(null, null, `/document/${createDocument.id}`);
+          createDocument.content = getItem(documentLocalSaveKey, "").content;
+
+          documentLocalSaveKey = `temp-document-${createDocument.id}`;
+
+          await request(`/documents/${createDocument.id}`, {
+            method: "PUT",
+            body: JSON.stringify({
+              title: createDocument.title,
+              content: createDocument.content,
+            }),
+          });
+
+          setItem(documentLocalSaveKey, {
```
documentLocalSaveKey 가 계속 변경되고 변경된 시점으로부터 멀리떨어진 곳에서 사용되다보니, 가독성이 약간 떨어지는것같아요..ㅜ😭

ienrum commented on src/components/DocumentEditSection.js:
```diff
@@ -0,0 +1,156 @@
+            documentId: createDocument.id,
+            document: createDocument,
+          });
+        } else {
```
if 블록이 너무 길다보니 else 조건이 무엇인지 잘 이해하기 어려웠어요.. if else 로 보완할수있을것 같아요!

ienrum commented on src/components/DocumentEditSection.js:
```diff
@@ -0,0 +1,156 @@
+        } else {
+          editor.setState(defaultValue);
+        }
+      } else {
+        if (this.state.documentId !== "new") {
+          await fetchDocument();
+          return;
+        }
```
여기도 if else 나 if 를 삭제할수도있을것같아요!

ienrum commented on src/components/DocumentEditSection.js:
```diff
@@ -0,0 +1,156 @@
+
+    if (documentId === "new") return;
+
+    const document = await request(`/documents/${documentId}`);
+
+    documentLocalSaveKey = `temp-document-${documentId}`;
+
+    const storedDocument = getItem(documentLocalSaveKey, {
+      title: "",
+      content: "",
+    });
+
+    const { tempSaveDate } = storedDocument;
+    if (tempSaveDate && tempSaveDate > document.updatedAt) {
```
여기도 document 선언부를 쓰는곳에 가깝게 붙여주면 좋을것 같아요!

ienrum commented on src/components/Editor.js:
```diff
@@ -0,0 +1,162 @@
+
+export default function Editor({ $target, initialState, onEditing }) {
+  const $editor = document.createElement("div");
+  $editor.setAttribute("class", "editor-wrap");
```
set class를 활용해도 좋을것같아요!

ienrum commented on src/components/Editor.js:
```diff
@@ -0,0 +1,162 @@
+  const $editor = document.createElement("div");
+  $editor.setAttribute("class", "editor-wrap");
+
+  $editor.innerHTML = `
+    <input type="text" id="editor-title" name="title" placeholder="제목을 입력해주세요." />
+    <div id="editor-content" contentEditable="true" style="width: 600px;"></div>
+  `;
```
this.render 함수 부분에 넣는 시도도 좋을것 같아요!😀

ienrum commented on src/components/Editor.js:
```diff
@@ -0,0 +1,162 @@
+    const elementPosition = getPosition(node.parentElement, offset);
+
+    const $empty = document.createElement("span");
+    $empty.setAttribute("class", "empty");
```
여기도 setClass 를 쓰시면 명화할것 같아요!

ienrum reviewed (commented):
수고 많으셨습니다!☺️

kimyou1102 committed:
refactor: 불필요한 let const로 변경

kimyou1102 committed:
refactor: 태그 존재 확인 시 반복되는 코드 함수화

kimyou1102 committed:
refactor: 코드 리뷰 반영

- if-else문 switch문으로 변경
- 불필요한 코드 제거
- editorSection태그 생성로직 일관되게 해당 컴포넌트 안에서 생성
- 선언부와 가깝게 변수 위치 변경
- 오타 수정
- if블록 길 경우 else를 else if로 변경

kimyou1102 committed:
refactor: history.replaceState 기능 함수화
